### PR TITLE
perf: cache HTTP route when looked up by filters

### PR DESCRIPTION
**Pull request by mattklein123**
If multiple filters consult the route table, the same route can be looked
up many times. With large route tables, this can become quite expensive.
This change adds caching for the route lookup and changes the interfaces
so that headers are no longer passed when getting a stable route. In
the future we may need to add route cache clearing, but for now this is
fine.

fixes https://github.com/lyft/envoy/issues/372

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @rshriram

File: source/common/http/filter/fault_filter.cc [hunk 1]
```diff
@@ -149,11 +149,11 @@ void FaultFilter::abortWithHTTPStatus() {
   callbacks_->requestInfo().setResponseFlag(Http::AccessLog::ResponseFlag::FaultInjected);
 }
 
-bool FaultFilter::matchesTargetCluster(const HeaderMap& headers) {
+bool FaultFilter::matchesTargetCluster(const HeaderMap&) {
```
Comment (rshriram) [inline source/common/http/filter/fault_filter.cc]: You could get rid of the headers arg completely. It's used only to obtain the target cluster from connection manager.

Comment (mattklein123) [inline source/common/http/filter/fault_filter.cc]: will fix

File: include/envoy/http/filter.h [hunk 2]
```diff
@@ -84,12 +84,16 @@ class StreamFilterCallbacks {
...
 
   /**
-   * @return const Router::StableRouteTable& the route table used for HTTP routing decisions.
-   *         The most important consumer of this information is the router filter. However,
-   *         other filters may wish to know the route that will be taken for various reasons so
-   *         the information is supplied to all filters. The table may be empty.
+   * Returns the route for the current request. The assumption is that the implementation can do
+   * caching where applicable to avoid multiple lookups.
+   *
+   * NOTE: This breaks down a bit if the caller knows it has modified something that would effect
```
Comment (junr03) [inline include/envoy/http/filter.h]: nit: affect

File: include/envoy/http/filter.h [hunk 3]
```diff
@@ -84,12 +84,16 @@ class StreamFilterCallbacks {
...
+   *       routing (such as the request headers). In practice, we don't do this anywhere currently,
+   *       but in the future we might need to provide the ability to clear the cache if a filter
+   *       knows that it has modified the headers in a way that would effect routing. In the future
```
Comment (junr03) [inline include/envoy/http/filter.h]: nit: affect

**Comment by mattklein123** [comment]
RDS does not require cache invalidation in and of itself, as the caching is per request, not global.

**Pull request merged**